File: src/bot/context-weave.ts
import type { LarkChannel } from '@larksuiteoapi/node-sdk';
import { log } from '../core/logger';

/**
 * Inbound CONTEXT weaving — give codex the上下文 a Feishu @ alone doesn't carry.
 *
 * Feishu's message event only delivers the triggering message itself. Two common
 * blind spots follow:
 *   1. 引用消息 (quote reply): the user 引用 another message + @bot. The quoted
 *      message id rides in `NormalizedMessage.replyToMessageId` (= the raw
 *      event's `parent_id`), but its CONTENT is never pushed — it must be pulled
 *      with `im.v1.message.get`.
 *   2. 话题上文 (topic history): people chat in a topic, then @bot mid-thread with
 *      little/no description. Those人对人 messages were never @ the bot, so the
 *      bridge dropped them at the @门 — codex has never seen them. They must be
 *      pulled with `im.v1.message.list` (container_id_type='thread').
 *
 * codex has no native "quoted message" / "chat history" input, so — exactly like
 * inbound files (see {@link import('./media').weaveFileManifest}) — we fold the
 * pulled text into the prompt as clearly-fenced blocks PREPENDED to the user's
 * actual line (which stays the instruction codex acts on).
 *
 * The bot reads these with its OWN tenant identity: it already holds
 * `im:message.group_msg` (read ALL group messages, see config/scopes.ts) and
 * already calls `im.v1.message.get` (media.ts) — no second identity / lark-cli
 * needed. Every fetch is best-effort: a deleted message, a missing scope, or a
 * timeout is logged and skipped, never thrown — a missing block must not break
 * the turn.
 *
 * {@link sanitizeContext} is the SINGLE sanitization boundary for this
 * uploader-controlled text (mirrors media.cleanFileName): it strips control
 * chars and clamps length right before the text is woven into the prompt.
 */

/** Max chars woven for one quoted message (collapsed to a single line). */
const QUOTE_MAX = 800;
/** Max chars per thread-history line (collapsed to a single line each). */
const LINE_MAX = 280;
/** Max thread messages woven (most-recent kept). */
const THREAD_WEAVE_MAX = 20;
/** How many thread messages to PULL (one page, newest-first) before filtering —
 * bot/system/empty messages drop out, so pull extra headroom over THREAD_WEAVE_MAX.
 * 50 is the Feishu page_size ceiling; older messages beyond it are intentionally
 * dropped (we only want recent context) and the truncation is logged. */
const THREAD_PAGE_SIZE = 50;

/** A message pulled for context (the quoted message, or one thread-history entry). */
export interface ContextMessage {
  messageId: string;
  /** Best-effort display name (sender_name, else `用户<id尾4>`, else 某人). */
  senderName: string;
  /** Plain text extracted from the message body (NOT yet sanitized). */
  text: string;
  /** sender_type === 'user' — false for the bot itself, other apps, system. */
  fromUser: boolean;
  /** create_time as epoch ms (0 when unknown). */
  createTime: number;
}

/** Subset of the `im.v1.message.get` / `.list` item shape we read. */
interface RawMsgItem {
  message_id?: string;
  msg_type?: string;
  create_time?: string;
  deleted?: boolean;
  sender?: { id?: string; sender_type?: string; sender_name?: string };
  body?: { content?: string };
  mentions?: { key?: string; name?: string }[];
}

// ── pulling ──────────────────────────────────────────────────────────────

/**
 * Fetch the quoted/replied message's content by id. Returns undefined when the
 * message is missing/deleted/empty or the call fails (all best-effort).
 */
export async function fetchQuotedMessage(
  channel: LarkChannel,
  messageId: string,
): Promise<ContextMessage | undefined> {
  try {
    const res = await channel.rawClient.im.v1.message.get({ path: { message_id: messageId } });
    const items = (res.data as { items?: RawMsgItem[] } | undefined)?.items ?? [];
    const item = items[0];
    if (!item || item.deleted) return undefined;
    const cm = toContextMessage(item);
    return cm.text.trim() ? cm : undefined;
  } catch (err) {
    log.warn('intake', 'quote-fetch-failed', { messageId, err: String(err) });
    return undefined;
  }
}

/**
 * Pull the human messages in a topic the bot hasn't fed to codex yet.
 *   - `sinceTime === 0` (fresh session): return the most-recent N人对人 messages
 *     as opening context.
 *   - `sinceTime > 0` (existing session): return only messages newer than that —
 *     the chatter that happened between bot turns (codex already has the rest).
 * Bot/app/system messages and the triggering @ message are filtered out. Returns
 * [] on any failure (best-effort). Result is oldest→newest.
 */
export async function fetchThreadContext(
  channel: LarkChannel,
  threadId: string,
  opts: { sinceTime?: number; excludeMessageId?: string; limit?: number } = {},
): Promise<ContextMessage[]> {
  const limit = opts.limit ?? THREAD_WEAVE_MAX;
  const since = opts.sinceTime ?? 0;
  try {
    const res = await channel.rawClient.im.v1.message.list({
      params: {
        container_id_type: 'thread',
        container_id: threadId,
        sort_type: 'ByCreateTimeDesc',
        page_size: THREAD_PAGE_SIZE,
      },
    });
    const items = (res.data as { items?: RawMsgItem[] } | undefined)?.items ?? [];
    const picked = items
      .filter((it) => !it.deleted)
      .map(toContextMessage)
      .filter(
        (m) =>
          m.fromUser && // drop the bot's own replies, other apps, system notices
          m.messageId !== opts.excludeMessageId && // drop the triggering @ message
          (since === 0 || m.createTime > since) && // delta only for existing sessions
          m.text.trim().length > 0,
      );
    // `list` returned newest-first; weave oldest→newest, keeping the most recent `limit`.
    picked.sort((a, b) => a.createTime - b.createTime);
    const out = picked.slice(-limit);
    if (picked.length > out.length) {
      log.info('intake', 'thread-context-truncated', { threadId, kept: out.length, total: picked.length });
    }
    return out;
  } catch (err) {
    log.warn('intake', 'thread-context-failed', { threadId, err: String(err) });
    return [];
  }
}

/**
 * Narrow a speculatively-pulled FULL history (`sinceTime: 0`) down to the delta
 * an existing session needs — byte-for-byte equivalent to having called
 * {@link fetchThreadContext} with `sinceTime` directly. The equivalence holds
 * because the API pull is identical for every sinceTime (one newest-first page;
 * sinceTime is purely a local filter) and watermark survivors are always the
 * newest messages, so they can never be evicted by the most-recent-`limit`
 * slice in favor of older ones. This is what lets the intake path pull 话题上文
 * in parallel with resolveThread instead of waiting for its codexEmpty answer.
 */
export function filterHistorySince(msgs: ContextMessage[], sinceTime: number): ContextMessage[] {
  if (sinceTime <= 0) return msgs;
  return msgs.filter((m) => m.createTime > sinceTime);
}

function toContextMessage(item: RawMsgItem): ContextMessage {
  const id = item.sender?.id ?? '';
  const name = item.sender?.sender_name || (id ? `用户${id.slice(-4)}` : '某人');
  return {
    messageId: item.message_id ?? '',
    senderName: name,
    text: extractMessageText(item.msg_type, item.body?.content, item.mentions),
    fromUser: item.sender?.sender_type === 'user',
    createTime: Number(item.create_time) || 0,
  };
}

// ── content → plain text ───────────────────────────────────────────────────

/**
 * Extract human-readable text from a raw message body. `text`/`post` yield their
 * words (with @mentions resolved to names); everything else collapses to a short
 * placeholder (`[图片]`, `[文件：a.log]`, …) so codex knows something was there
 * without us trying to render it. Exported for testing — the body shapes are the
 * bug-prone part. Returned text is NOT yet sanitized (caller's job).
 */
export function extractMessageText(
  msgType: string | undefined,
  content: string | undefined,
  mentions?: { key?: string; name?: string }[],
): string {
  if (!content) return placeholderFor(msgType);
  let parsed: unknown;
  try {
    parsed = JSON.parse(content);
  } catch {
    return placeholderFor(msgType);
  }
  switch (msgType) {
    case 'text':
      return replaceMentions((parsed as { text?: string } | null)?.text ?? '', mentions);
    case 'post':
      return replaceMentions(extractPostText(parsed), mentions);
    case 'image':
      return '[图片]';
    case 'audio':
      return '[语音]';
    case 'media':
      return '[视频]';
    case 'file': {
      const name = (parsed as { file_name?: string } | null)?.file_name;
      return name ? `[文件：${name}]` : '[文件]';
    }
    case 'sticker':
      return '[表情]';
    case 'interactive':
      return '[卡片消息]';
    case 'share_chat':
      return '[分享群名片]';
    case 'share_user':
      return '[分享个人名片]';
    case 'merge_forward':
    case 'forward':
      return '[合并转发消息]';
    default:
      return placeholderFor(msgType);
  }
}

function placeholderFor(msgType: string | undefined): string {
  return msgType ? `[${msgType} 消息]` : '[消息]';
}

/** Pull text out of a `post` (rich-text) body — both the direct
 * `{title, content:[[node]]}` shape and the locale-wrapped `{zh_cn:{...}}` one. */
function extractPostText(parsed: unknown): string {
  if (!parsed || typeof parsed !== 'object') return '';
  const obj = parsed as Record<string, unknown>;
  let title = obj.title;
  let blocks = obj.content;
  if (!Array.isArray(blocks)) {
    for (const v of Object.values(obj)) {
      if (v && typeof v === 'object' && Array.isArray((v as { content?: unknown }).content)) {
        title = (v as { title?: unknown }).title;
        blocks = (v as { content?: unknown }).content;
        break;
      }
    }
  }
  const parts: string[] = [];
  if (typeof title === 'string' && title.trim()) parts.push(title.trim());
  if (Array.isArray(blocks)) {
    for (const line of blocks) {
      if (!Array.isArray(line)) continue;
      const lineText = line.map(nodeToText).join('');
      if (lineText) parts.push(lineText);
    }
  }
  return parts.join('\n');
}

function nodeToText(node: unknown): string {
  if (!node || typeof node !== 'object') return '';
  const n = node as Record<string, unknown>;
  switch (n.tag) {
    case 'text':
      return typeof n.text === 'string' ? n.text : '';
    case 'a':
      return typeof n.text === 'string' ? n.text : typeof n.href === 'string' ? n.href : '';
    case 'at': {
      const name = typeof n.user_name === 'string' ? n.user_name : '';
      return name ? `@${name}` : '@某人';
    }
    case 'img':
      return '[图片]';
    case 'media':
      return '[视频]';
    case 'emotion':
      return '[表情]';
    default:
      return typeof n.text === 'string' ? n.text : '';
  }
}

/** Replace `@_user_N` / `@_all` placeholder tokens in a `text` body with the
 * mentioned names so codex reads "@张三" instead of an opaque key. */
function replaceMentions(text: string, mentions?: { key?: string; name?: string }[]): string {
  if (!text || !mentions?.length) return text;
  let out = text;
  for (const m of mentions) {
    if (!m.key) continue;
    out = out.split(m.key).join(m.name ? `@${m.name}` : '@某人');
  }
  return out;
}

// ── sanitize + weave ─────────────────────────────────────────────────────

/**
 * The single sanitization boundary for pulled, uploader-controlled context text
 * (mirrors media.cleanFileName). Strips control chars, normalizes newlines, and
 * clamps length. `oneLine` collapses ALL whitespace to single spaces — both
 * weave callers use it so a crafted body can't forge fenced-block structure
 * (a `\n]\n[fake]`) to inject instructions. The multi-line branch (kept for
 * generality) only squeezes whitespace runs. Exported for testing.
 */
export function sanitizeContext(s: string, maxLen: number, oneLine: boolean): string {
  if (!s) return '';
  let out = s
    // eslint-disable-next-line no-control-regex
    .replace(/[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]/g, '') // drop control chars (keep \t \n \r)
    .replace(/\r\n?/g, '\n');
  out = oneLine ? out.replace(/\s+/g, ' ') : out.replace(/[ \t]+\n/g, '\n').replace(/\n{3,}/g, '\n\n');
  out = out.trim();
  return out.length > maxLen ? `${out.slice(0, maxLen)}…` : out;
}

/**
 * Prepend a fenced "引用的消息" block before the user's text. Returns `text`
 * unchanged when there's no quoted message or it sanitizes to empty.
 */
export function weaveQuote(text: string, quoted: ContextMessage | undefined): string {
  if (!quoted) return text;
  const who = sanitizeContext(quoted.senderName, 40, true) || '某人';
  // oneLine=true: collapse newlines so a crafted quote body (or @mention name)
  // can't forge a `\n]\n[fake block]` to fabricate context — same defense as
  // weaveThreadHistory. sanitizeContext is the single boundary for this text.
  const body = sanitizeContext(quoted.text, QUOTE_MAX, true);
  if (!body) return text;
  const block = `[用户引用了一条消息（来自 ${who}）：\n${body}\n]`;
  const base = text.trim();
  return base ? `${block}\n\n${base}` : block;
}

/**
 * Prepend a fenced "话题上文" block (one line per message, time-ascending) before
 * the user's text. Returns `text` unchanged when there are no messages.
 */
export function weaveThreadHistory(text: string, msgs: ContextMessage[]): string {
  if (msgs.length === 0) return text;
  const lines = msgs
    .map((m) => {
      const who = sanitizeContext(m.senderName, 40, true) || '某人';
      const body = sanitizeContext(m.text, LINE_MAX, true);
      return body ? `${who}：${body}` : '';
    })
    .filter((l) => l.length > 0);
  if (lines.length === 0) return text;
  const block = `[话题中在此之前已有的消息（按时间先后排列，供你理解上下文）：\n${lines.join('\n')}\n]`;
  const base = text.trim();
  return base ? `${block}\n\n${base}` : block;
}

File: internal/bot/context_weave.go
package bot

// context_weave.go —— 入站消息上下文织入的纯函数（对齐 TS bot/context-weave）。
// extractMessageText（消息体→文本）+ sanitizeContext（抗注入）+ weaveQuote/weaveThreadHistory/weaveSender（织入 prompt）。
// 飞书 SDK 部分（fetchQuotedMessage/fetchThreadContext）后续 bot/feishu wrapper port 后接上。

import (
	"encoding/json"
	"regexp"
	"strings"
)

// ContextMessage 拉取的上下文消息（引用消息 / 话题历史条目）。
type ContextMessage struct {
	MessageID  string
	SenderName string
	Text       string
	FromUser   bool
	CreateTime int64 // epoch ms
}

// Mention @mention（key=@_user_N, name=展示名）。
type Mention struct{ Key, Name string }

const (
	quoteMax       = 800
	lineMax        = 280
	threadWeaveMax = 20
	senderNameMax  = 40
)

const cardUpgradeHint = "请升级至最新版本客户端"

// ExtractMessageText 从原始消息体提取人类可读文本。text/post 提文字（@mention 解析），
// 其余折叠为占位符（[图片]/[文件：x]/…）。导出供测试。
func ExtractMessageText(msgType, content string, mentions []Mention) string {
	if content == "" {
		return placeholderFor(msgType)
	}
	var parsed map[string]any
	if err := json.Unmarshal([]byte(content), &parsed); err != nil {
		return placeholderFor(msgType)
	}
	switch msgType {
	case "text":
		t, _ := parsed["text"].(string)
		return replaceMentions(t, mentions)
	case "post":
		return replaceMentions(extractPostText(parsed), mentions)
	case "image":
		return "[图片]"
	case "audio":
		return "[语音]"
	case "media":
		return "[视频]"
	case "file":
		if name, ok := parsed["file_name"].(string); ok && name != "" {
			return "[文件：" + name + "]"
		}
		return "[文件]"
	case "sticker":
		return "[表情]"
	case "interactive":
		if t := ExtractCardText(parsed); t != "" {
			return t
		}
		return "[卡片消息]"
	case "share_chat":
		return "[分享群名片]"
	case "share_user":
		return "[分享个人名片]"
	case "merge_forward", "forward":
		return "[合并转发消息]"
	}
	return placeholderFor(msgType)
}

func placeholderFor(msgType string) string {
	if msgType == "" {
		return "[消息]"
	}
	return "[" + msgType + " 消息]"
}

// extractPostText 从 post（富文本）body 提取文本（支持 {title,content} 和 locale-wrapped {zh_cn:{...}}）。
func extractPostText(parsed map[string]any) string {
	title, _ := parsed["title"].(string)
	blocks, _ := parsed["content"].([]any)
	// locale-wrapped 形态。
	if blocks == nil {
		for _, v := range parsed {
			if vm, ok := v.(map[string]any); ok {
				if c, ok := vm["content"].([]any); ok {
					if t, ok := vm["title"].(string); ok {
						title = t
					}
					blocks = c
					break
				}
			}
		}
	}
	var parts []string
	if strings.TrimSpace(title) != "" {
		parts = append(parts, strings.TrimSpace(title))
	}
	for _, line := range blocks {
		lineArr, ok := line.([]any)
		if !ok {
			continue
		}
		var lineText strings.Builder
		for _, node := range lineArr {
			lineText.WriteString(nodeToText(node))
		}
		if lineText.Len() > 0 {
			parts = append(parts, lineText.String())
		}
	}
	return strings.Join(parts, "\n")
}

func nodeToText(node any) string {
	n, ok := node.(map[string]any)
	if !ok {
		return ""
	}
	tag, _ := n["tag"].(string)
	switch tag {
	case "text":
		t, _ := n["text"].(string)
		return t
	case "a":
		t, _ := n["text"].(string)
		if t != "" {
			return t
		}
		href, _ := n["href"].(string)
		return href
	case "at":
		name, _ := n["user_name"].(string)
		if name != "" {
			return "@" + name
		}
		return "@某人"
	case "img":
		return "[图片]"
	case "media":
		return "[视频]"
	case "emotion":
		return "[表情]"
	}
	t, _ := n["text"].(string)
	return t
}

// ExtractCardText 从 interactive（卡片）body 提取文本（down-convert 形态 {title,elements:[[node]]}）。
func ExtractCardText(parsed map[string]any) string {
	var parts []string
	title := textValue(parsed["title"])
	if strings.TrimSpace(title) != "" {
		parts = append(parts, strings.TrimSpace(title))
	}
	if elements, ok := parsed["elements"].([]any); ok {
		for _, line := range elements {
			var nodes []any
			if arr, ok := line.([]any); ok {
				nodes = arr
			} else {
				nodes = []any{line}
			}
			var lineText strings.Builder
			for _, n := range nodes {
				lineText.WriteString(cardNodeToText(n))
			}
			lt := strings.TrimSpace(lineText.String())
			if lt != "" && !strings.Contains(lt, cardUpgradeHint) {
				parts = append(parts, lt)
			}
		}
	}
	return strings.TrimSpace(strings.Join(parts, "\n"))
}

func textValue(t any) string {
	if s, ok := t.(string); ok {
		return s
	}
	if m, ok := t.(map[string]any); ok {
		if c, ok := m["content"].(string); ok {
			return c
		}
	}
	return ""
}

func cardNodeToText(node any) string {
	if s, ok := node.(string); ok {
		return s
	}
	n, ok := node.(map[string]any)
	if !ok {
		return ""
	}
	tag, _ := n["tag"].(string)
	switch tag {
	case "text":
		return textValue(n["text"])
	case "a":
		t := textValue(n["text"])
		if t != "" {
			return t
		}
		href, _ := n["href"].(string)
		return href
	case "at":
		name, _ := n["user_name"].(string)
		if name != "" {
			return "@" + name
		}
		return "@某人"
	case "note":
		if els, ok := n["elements"].([]any); ok {
			var sb strings.Builder
			for _, e := range els {
				sb.WriteString(cardNodeToText(e))
			}
			return sb.String()
		}
		return ""
	case "button":
		label := textValue(n["text"])
		if label != "" {
			return "[按钮：" + label + "]"
		}
		return ""
	case "img":
		return ""
	}
	return textValue(n["text"])
}

func replaceMentions(text string, mentions []Mention) string {
	if text == "" || len(mentions) == 0 {
		return text
	}
	out := text
	for _, m := range mentions {
		if m.Key == "" {
			continue
		}
		name := "@某人"
		if m.Name != "" {
			name = "@" + m.Name
		}
		out = strings.ReplaceAll(out, m.Key, name)
	}
	return out
}

var (
	controlCharRe = regexp.MustCompile(`[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]`)
	crRe          = regexp.MustCompile(`\r\n?`)
	whiteAllRe    = regexp.MustCompile(`\s+`)
	trailNLRe     = regexp.MustCompile(`[ \t]+\n`)
	multiNLRe     = regexp.MustCompile(`\n{3,}`)
)

// SanitizeContext 唯一 sanitize 边界（uplink-controlled text 织入 prompt）。
// 剥控制字符 + 归一换行 + clamp 长度。oneLine=true 折叠所有空白（防伪造 fenced block）。
func SanitizeContext(s string, maxLen int, oneLine bool) string {
	if s == "" {
		return ""
	}
	out := controlCharRe.ReplaceAllString(s, "")
	out = crRe.ReplaceAllString(out, "\n")
	if oneLine {
		out = whiteAllRe.ReplaceAllString(out, " ")
	} else {
		out = trailNLRe.ReplaceAllString(out, "\n")
		out = multiNLRe.ReplaceAllString(out, "\n\n")
	}
	out = strings.TrimSpace(out)
	runes := []rune(out)
	if len(runes) > maxLen {
		return string(runes[:maxLen]) + "…"
	}
	return out
}

func prependBlock(block, text string) string {
	base := strings.TrimSpace(text)
	if base == "" {
		return block
	}
	return block + "\n\n" + base
}

// WeaveQuote 在用户文本前织入引用消息块（oneLine 抗注入）。
func WeaveQuote(text string, quoted *ContextMessage) string {
	if quoted == nil {
		return text
	}
	who := SanitizeContext(quoted.SenderName, 40, true)
	if who == "" {
		who = "某人"
	}
	body := SanitizeContext(quoted.Text, quoteMax, true)
	if body == "" {
		return text
	}
	block := "[用户引用了一条消息（来自 " + who + "）：\n" + body + "\n]"
	return prependBlock(block, text)
}

// WeaveThreadHistory 在用户文本前织入话题上文块（一行一条，时间升序）。
func WeaveThreadHistory(text string, msgs []ContextMessage) string {
	if len(msgs) == 0 {
		return text
	}
	var lines []string
	for _, m := range msgs {
		who := SanitizeContext(m.SenderName, 40, true)
		if who == "" {
			who = "某人"
		}
		body := SanitizeContext(m.Text, lineMax, true)
		if body != "" {
			lines = append(lines, who+"："+body)
		}
	}
	if len(lines) == 0 {
		return text
	}
	block := "[话题中在此之前已有的消息（按时间先后排列，供你理解上下文）：\n" + strings.Join(lines, "\n") + "\n]"
	return prependBlock(block, text)
}

// WeaveSender 在用户文本前织入发信人身份块。
func WeaveSender(text, senderID, senderName string) string {
	id := strings.TrimSpace(senderID)
	if id == "" {
		return text
	}
	who := SanitizeContext(senderName, senderNameMax, true)
	if who == "" {
		who = "某用户"
	}
	block := "[本条消息的发信人：" + who + "（open_id：" + id + "）]"
	return prependBlock(block, text)
}

// FilterHistorySince 把全量历史按 sinceTime 过滤为增量。
func FilterHistorySince(msgs []ContextMessage, sinceTime int64) []ContextMessage {
	if sinceTime <= 0 {
		return msgs
	}
	var out []ContextMessage
	for _, m := range msgs {
		if m.CreateTime > sinceTime {
			out = append(out, m)
		}
	}
	return out
}
